### PR TITLE
feat(dataset): external storage backend

### LINKED ISSUE #3279: Store file sizes for datasets with cloud storage
`renku dataset ls-files` command retrieves file sizes on-the-fly. However, for files in a cloud storage, it's not possible/easy to do that since the dataset might not be mounted at all.
We should store file sizes when adding files to a dataset so that we don't need to calculate them each time.

### LINKED ISSUE #3284: Support external files as a storage backend
`renku dataset add` command allowing adding data from the local filesystem as `external` by passing the `--external` flag. External data is not copied to the project and therefore not tracked in the git. Instead, a symlink to the source data is created and used when dealing with data.

External data can be added to each dataset along with local data. This is confusing for users since it's not clear which data is external and which isn't. Moreover, the source of the data isn't clear and users need to follow symlinks to see where the data comes from. Finally, since there is a symlink to the data it's easy to corrupt the source data.

We can improve UX around the external data by supporting a local storage backend that acts similar to external data but at a dataset level (storage backends are per dataset). This would eliminate the confusion between external and local data. Also, the source of data is clearly visible. We can avoid corrupting source data by using a copy of data (`renku dataset pull`) or by mounting data as read-only.

We should remove the current support for external individual data.

### PR DESCRIPTION
**Pull request by mohammad-alisafaee**
# Description

Removes external files. Replaces `is_external` with `linked` for linked files. Adds an `external` storage backend for external files. Store file sizes when adding files to datasets.

## TODO
* [x] Set minimum version to `2.4.0` since we add new fields to the metadata.

Fixes #3284 
Fixes #3279

### CODE REVIEW

File: renku/core/dataset/providers/api.py [hunk 1]
```diff
@@ -26,6 +26,11 @@
 from renku.core.util.util import NO_VALUE, NoValueType
 from renku.domain_model.dataset_provider import IDatasetProviderPlugin
 
+try:
+    from typing_extensions import Protocol  # NOTE: Required for Python 3.7 compatibility
```
Comment (Panaetius) [inline renku/core/dataset/providers/api.py:30]: We don't support 3.7 anymore

**Review by Panaetius — CHANGES_REQUESTED** [review]
should we separate renku.core.dataset.providers to make it clear what is for storage and what is for just add? It's getting a bit crowded in there.

And I'm a bit torn on the internal naming going from `external` to `cloud`, since it's not all cloud-based now. That could easily lead to misunderstandings. I like `external` or `remote` more

File: renku/command/checks/datasets.py [hunk 2]
```diff
@@ -194,3 +196,44 @@ def check_dataset_files_outside_datadir(fix, dataset_gateway: IDatasetGateway, *
...
+        external_files_str = "\n\t".join(sorted(external_files))
+        problems = (
+            f"\n{WARNING}: External files are deprecated in favor of an external dataset backend.\n"
+            "Use 'renku dataset rm' or rerun 'renku doctor' with '--fix' flag to remove them:\n\t"
+            f"{external_files_str}\n"
+        )
+        return False, problems
+
+    for name, files in datasets.items():
+        file_unlink(name=name, yes=True, dataset_files=files)
```
[the remark this answers is not in the document]
Comment (mohammad-alisafaee) [reply renku/command/checks/datasets.py:237]: I believe there aren't many uses of external data (if any at all) outside our test projects. It's not worth it to do an automatic migration. We don't have a document yet for the external storages. I'll add this once we write the document.

**Comment by mohammad-alisafaee** [comment]
> should we separate renku.core.dataset.providers to make it clear what is for storage and what is for just add? It's getting a bit crowded in there.

There are providers that are both storage and for add (S3, Azure, ...). I'm not sure how we can address them. I believe `renku.core.dataset.providers` isn't the best place for providers anymore. We should remove them outside the `dataset` module and put them directly under `core`. Maybe also rename them to data-provider (since we use provider in other places).

> And I'm a bit torn on the internal naming going from `external` to `cloud`, since it's not all cloud-based now. That could easily lead to misunderstandings. I like `external` or `remote` more

I changed it to cloud since I've noticed people are using this term. `Remote` makes more sense although it's a bit overused. Let's discuss this in the meeting tomorrow to decide on the terminology.

File: renku/command/checks/datasets.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (Panaetius) [reply renku/command/checks/datasets.py:237]: At least a sentence that it should be re-added after this would make sense to me.

**Review by Panaetius — APPROVED** [review]

**Review by Panaetius — APPROVED** [review]

**Pull request merged**